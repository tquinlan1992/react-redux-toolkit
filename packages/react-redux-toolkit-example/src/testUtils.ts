import {
  createConnectHooksTester,
  createConnectReactTester,
  createConnectReduxTester,
  mockHooks,
} from '@tquinlan1992/react-redux-toolkit';
import { extraArgsConnectProps, mappedStoreActions } from './connectProps';

import { RootState } from './store';

const testUtilsExtraArgs = {
  ...extraArgsConnectProps,
  extraArgsMapHooks: {
    mockHooks: mockHooks(extraArgsConnectProps.extraArgsHooks.hooks),
  },
  actions: mappedStoreActions,
};

export const testComponent = createConnectReduxTester<RootState, typeof testUtilsExtraArgs>({
  ...testUtilsExtraArgs,
});

export const testConnectHooks =
  createConnectHooksTester<typeof testUtilsExtraArgs>(testUtilsExtraArgs);

export const testConnectReact =
  createConnectReactTester<typeof testUtilsExtraArgs>(testUtilsExtraArgs);
