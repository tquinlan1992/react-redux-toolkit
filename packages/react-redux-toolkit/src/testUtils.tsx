import { omit } from 'lodash';
import { UseFormReturn } from 'react-hook-form';
import ShallowRenderer from 'react-test-renderer/shallow';
import * as Yup from 'yup';

/* eslint-disable @typescript-eslint/no-explicit-any */
import { MutationTrigger } from '@reduxjs/toolkit/dist/query/react/buildHooks';

const isTestProp = (prop: string) =>
  [
    '$$typeof',
    'nodeType',
    'tagName',
    'hasAttribute',
    '@@__IMMUTABLE_ITERABLE__@@',
    '@@__IMMUTABLE_RECORD__@@',
    '_isMockFunction',
    'Symbol(Symbol.toStringTag)',
    'toJSON',
    '@@__IMMUTABLE_ITERABLE__@@',
    '',
  ].includes(prop);

export type DeepPartial<T> = {
  [P in keyof T]?: DeepPartial<T[P]>;
};

export function getPartialDeep<Value>(value: DeepPartial<Value>): Value {
  return value as any;
}

const errorsHandler = {
  get: function (target: Record<string, any>, prop: string) {
    return {
      message: `${prop} errors`,
    };
  },
};

const mockForm = {
  register: jest.fn((name: string) => ({ name, id: name })),
  handleSubmit: jest.fn(),
  formState: {
    errors: new Proxy({}, errorsHandler),
  },
} as any as UseFormReturn<any>;

const mockActions = new Proxy(
  {},
  {
    get: function (target: Record<string, any>, prop: string) {
      return `actions.${prop}`;
    },
  },
);

const hookReturnProxy = (hookName: string) =>
  new Proxy(
    {},
    {
      get: function (target: Record<string, any>, prop: string) {
        return `${hookName}.${prop}`;
      },
    },
  );

export function mockHooks<Hooks extends Record<string, any>>(hooks: Hooks): Hooks {
  const cache: Record<string, any> = {};
  return new Proxy(
    {},
    {
      get: function (target: Record<string, any>, prop: string) {
        const cached = cache[prop];
        if (cached) {
          return cached;
        } else {
          cache[prop] = jest.fn(() => hookReturnProxy(prop));
          return cache[prop];
        }
      },
    },
  ) as any;
}

const proxyData = () =>
  new Proxy(
    { data: 'apiQueries.data' },
    {
      get: function (target: Record<string, any>, prop: string) {
        if (target[prop] === undefined && !isTestProp(prop)) {
          try {
            return `data.${prop}`;
          } catch {
            return target[prop];
          }
        } else {
          return target[prop];
        }
      },
    },
  );

export const getMockApiResult = (rootProp: string) => {
  const error = { error: { name: `${rootProp}.error` } };
  const data = true;
  const isLoading = `${rootProp}.isLoading`;
  const initialValue: any = { isLoading };
  if (error) {
    initialValue.error = error;
  }
  if (data) {
    initialValue.data = proxyData();
  }
  return new Proxy(initialValue, {
    get: (target: Record<string, any>, prop: string) => {
      if (prop === 'isLoading') {
        return isLoading;
      }
      if (prop === 'error' && error) {
        return error;
      }
      if (!isTestProp(prop)) {
        try {
          if (prop.startsWith('data') && data) {
            return target[prop];
          }
          if (prop.startsWith('data') && !data) {
            return undefined;
          }
          if (prop === 'isLoading') {
            return false;
          }
          return `apiQueries.${prop}`;
        } catch {
          return target[prop];
        }
      }
    },
  });
};

const getMockApiQueries = () =>
  new Proxy(
    {},
    { get: (target: Record<string, any>, prop: string) => getMockApiResult(`apiQueries.${prop}`) },
  ) as any;

export const getMockQueryMutation = (rootProp: string) => {
  const mutationTrigger = `${rootProp}.mutationTrigger` as any as MutationTrigger<any>;
  return [mutationTrigger, getMockApiResult(rootProp)];
};

const getMockApiMutations = () =>
  new Proxy(
    {},
    {
      get: (target: Record<string, any>, prop: string) =>
        getMockQueryMutation(`apiMutations.${prop}`),
    },
  ) as any;

export function createTestComponent<
  AppState,
  Options extends {
    extraArgsMapState?: Record<string, any>;
    extraArgsComponent?: Record<string, any>;
    extraArgsMapHooks?: Record<string, any>;
  },
>({
  extraArgsMapState,
  extraArgsComponent,
  extraArgsMapHooks,
}: Options): <
  RRTKComponent extends {
    mapStateToProps: (
      args: { state: AppState; ownProps: any } & Options['extraArgsMapState'],
    ) => any;
    Component: <
      Args extends {
        actions: any;
        state: any;
        hooks: any;
        form: any;
        apiQueries: any;
        apiMutations: any;
      },
    >(
      args: Args,
    ) => any;
    mapHooks: (args: any) => any;
  },
>(
  componentName: string,
  rRTKComponent: RRTKComponent,
  {
    mapState,
    mapDispatch,
    component,
  }: {
    mapState: (
      args: {
        testMapState: (
          given: {
            ownProps?: Parameters<RRTKComponent['mapStateToProps']>[0]['ownProps'];
          } & DeepPartial<AppState>,
          expect: ReturnType<typeof rRTKComponent['mapStateToProps']>,
        ) => void;
      } & Options['extraArgsMapState'],
    ) => void;
    mapDispatch?: () => void;
    mapHooks: (
      args: Options['extraArgsMapHooks'] & {
        testMapHooks: (args: {
          given: Omit<Parameters<RRTKComponent['mapHooks']>[0], 'actions' | 'Yup'>;
          expect: {
            formDefaultValues?: ReturnType<RRTKComponent['mapHooks']>['defaultValues'];
          } & Omit<
            ReturnType<RRTKComponent['mapHooks']>,
            'hooks' | 'form' | 'actions' | 'apiQueries' | 'apiMutations'
          >;
          calledWith: () => {
            hooks?: {
              [P in keyof ReturnType<RRTKComponent['mapHooks']>['hooks']]: void;
            };
            apiQueries?: {
              [P in keyof ReturnType<RRTKComponent['mapHooks']>['apiQueries']]: void;
            };
            apiMutations?: {
              [P in keyof ReturnType<RRTKComponent['mapHooks']>['apiMutations']]: void;
            };
          };
        }) => void;
        testHookCalled: <Hook extends (args: any) => any>(
          hook: Hook,
          ...args: Parameters<Hook>
        ) => void;
      },
    ) => void;
    component: (
      args: Options['extraArgsComponent'] & {
        shallow: ShallowRenderer.ShallowRenderer;
        getPartialDeep: typeof getPartialDeep;
        testComponent: (
          props: DeepPartial<
            Omit<
              React.ComponentProps<typeof rRTKComponent['Component']>,
              'actions' | 'form' | 'apiQueries' | 'apiMutations'
            >
          >,
        ) => void;
      },
    ) => void;
  },
) => void {
  return (
    componentName,
    { mapStateToProps, mapHooks, Component },
    { mapState, mapDispatch, mapHooks: mapHooksTestProp, component },
  ) => {
    describe(componentName, () => {
      describe('mapState', () => {
        const testMapState = (given: any, expectValue: any) => {
          expect(
            mapStateToProps({
              state: omit(given, 'ownProps'),
              ownProps: given.ownProps,
              ...extraArgsMapState,
            }),
          ).toEqual(expectValue);
        };
        mapState({ testMapState });
      });
      describe('mapDispatch', () => {
        mapDispatch && mapDispatch();
      });
      describe('mapHooks', () => {
        const actions = { actions: mockActions };
        const testMapHooks = ({
          given,
          expect: expectValue = {},
          calledWith,
        }: {
          given: any;
          expect: any;
          calledWith: any;
        }) => {
          const mapHooksReturn = mapHooks({
            ...given,
            ...extraArgsMapState,
            Yup,
            actions: mockActions,
          });
          const returnWithoutForm = omit(
            mapHooksReturn,
            'hooks',
            'form',
            'apiQueries',
            'apiMutations',
          );
          expect({
            ...returnWithoutForm,
            ...(mapHooksReturn.form
              ? { formDefaultValues: mapHooksReturn.form.defaultValues }
              : {}),
          }).toEqual(expectValue);
          calledWith();
        };
        mapHooksTestProp &&
          mapHooksTestProp({
            testMapHooks,
            testHookCalled: (hook, ...args) => {
              expect(hook).toHaveBeenCalledTimes(1);
              expect(hook).toHaveBeenCalledWith(...args);
            },
            ...extraArgsMapHooks,
            ...actions,
          });
      });
      describe('renders', () => {
        const renderer = ShallowRenderer.createRenderer();
        const testComponent = (props: any) => {
          renderer.render(
            <Component
              {...props}
              form={mockForm}
              actions={mockActions}
              apiQueries={getMockApiQueries()}
              apiMutations={getMockApiMutations()}
            />,
          );
          expect(renderer.getRenderOutput()).toMatchSnapshot();
        };
        component({
          ...extraArgsComponent,
          shallow: renderer,
          testComponent,
          getPartialDeep,
        });
      });
    });
  };
}
