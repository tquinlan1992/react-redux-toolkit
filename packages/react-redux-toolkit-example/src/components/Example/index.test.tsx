import Example, {
  ConnectedHooksComponent,
  ConnectedReactComponent,
  ConnectedReactComponentOwnProps,
  ConnectedReactFCComponent,
  ConnectedReactFCComponentOwnProps,
} from '.';
import { testComponent, testConnectHooks, testConnectReact } from '../../testUtils';

testComponent('Example', Example, {
  mapState: ({ testMapState }) => {
    it('should mapstate', () => {
      testMapState(
        {
          example: {
            input1: 'input1',
            input2: 'input2',
          },
        },
        {
          input1: 'input1',
          input2: 'input2',
          hookInput: 'hookInput',
        },
      );
    });
  },
  mapHooks: ({ testMapHooks, testHookCalled, mockHooks }) => {
    it('should map hooks', () => {
      const state = { hookInput: 'hookInput', input1: 'input1', input2: 'input2' };
      testMapHooks({
        given: {
          state: state,
          hooks: mockHooks,
          ownProps: { test: 'test' },
        },
        expect: {
          state,
          formDefaultValues: {
            username: 'initial username',
          },
        },
        calledWith: () => ({
          hooks: {
            hook1: testHookCalled(mockHooks.hook1, 'hookInput'),
          },
        }),
      });
    });
  },
  component: ({ testComponent }) => {
    it('props 1', () => {
      testComponent({
        state: {
          input1: 'input1',
          input2: 'input2',
          hookInput: 'hookInput',
        },
        hooks: {
          hook1: 'hook1',
        },
        test: 'test',
      });
    });
  },
});

testConnectHooks('ConnectedHooksComponent', ConnectedHooksComponent, {
  mapHooks: ({ testMapHooks, testHookCalled, mockHooks }) => {
    it('should map hooks', () => {
      testMapHooks({
        given: {
          ownProps: {},
          hooks: mockHooks,
        },
        calledWith: () => ({
          hooks: {
            hook1: testHookCalled(mockHooks.hook1, 'testHookInput'),
          },
        }),
      });
    });
  },
  component: ({ testComponent }) => {
    it('props 1', () => {
      testComponent({
        hooks: {
          hook1: 'hook1',
        },
      });
    });
  },
});

testConnectReact('ConnectedReactComponent', ConnectedReactComponent, {
  component: ({ testComponent }) => {
    it('props 1', () => {
      testComponent({});
    });
  },
});

testConnectReact('ConnectedReactComponentOwnProps', ConnectedReactComponentOwnProps, {
  component: ({ testComponent }) => {
    it('props 1', () => {
      testComponent({ label: 'ConnectedReactComponentOwnProps Label' });
    });
  },
});

testConnectReact('ConnectedReactFCComponent', ConnectedReactFCComponent, {
  component: ({ testComponent }) => {
    it('props 1', () => {
      testComponent({});
    });
  },
});

testConnectReact('ConnectedReactFCComponentOwnProps', ConnectedReactFCComponentOwnProps, {
  component: ({ testComponent }) => {
    it('props 1', () => {
      testComponent({ label: 'ConnectedReactFCComponentOwnProps Label' });
    });
  },
});
