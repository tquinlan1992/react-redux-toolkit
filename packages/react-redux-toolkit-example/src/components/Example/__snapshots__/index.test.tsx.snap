// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConnectedHooksComponent renders props 1 1`] = `
<ForwardRef(Button)
  onClick={[Function]}
>
  hook1
</ForwardRef(Button)>
`;

exports[`ConnectedReactComponent renders props 1 1`] = `
<ForwardRef(Button)>
  Connect React Component
</ForwardRef(Button)>
`;

exports[`ConnectedReactComponentOwnProps renders props 1 1`] = `
<ForwardRef(Button)>
  ConnectedReactComponentOwnProps Label
</ForwardRef(Button)>
`;

exports[`ConnectedReactFCComponent renders props 1 1`] = `
<ForwardRef(Button)>
  ConnectedReactFCComponent Component
</ForwardRef(Button)>
`;

exports[`ConnectedReactFCComponentOwnProps renders props 1 1`] = `
<ForwardRef(Button)>
  ConnectedReactFCComponentOwnProps Label
</ForwardRef(Button)>
`;

exports[`Example renders props 1 1`] = `
<React.Fragment>
  <h1>
    Example of GreenXBlack redux toolkit
  </h1>
  <h1>
    Hooks results: 
    hook1
  </h1>
  <p>
    username input error: 
    username errors
  </p>
  <ForwardRef(Input)
    id="username"
    name="username"
    placeholder="username"
  />
  <ForwardRef(Button)>
    Handle React Form Submit
  </ForwardRef(Button)>
  <Memo(NoParams) />
  <ForwardRef(Button)
    onClick={[Function]}
  >
    Components.Button
  </ForwardRef(Button)>
  <div>
    <ForwardRef(Input)
      onChange={[Function]}
      placeholder="input1"
      value="input1"
    />
  </div>
  <div>
    <ForwardRef(Input)
      onChange={[Function]}
      placeholder="input2"
      value="input2"
    />
  </div>
  <div>
    <ForwardRef(Button)
      onClick={[MockFunction]}
    >
      Reset All Input
    </ForwardRef(Button)>
    <ForwardRef(Button)
      onClick={[Function]}
    >
      Reset Input 1
    </ForwardRef(Button)>
    <ForwardRef(Button)
      onClick={[Function]}
    >
      Reset Input 2
    </ForwardRef(Button)>
    <ForwardRef(Button)
      onClick={[Function]}
    >
      Set Input 1
    </ForwardRef(Button)>
    <ForwardRef(Button)
      onClick={[Function]}
    >
      Set Input 1
    </ForwardRef(Button)>
    <ForwardRef(Button)
      onClick={[Function]}
    >
      Set All Input
    </ForwardRef(Button)>
    <ForwardRef(Button)
      onClick={[MockFunction]}
    >
      Set From Thunk
    </ForwardRef(Button)>
    <Connected Hooks Component />
  </div>
  <ConnectReactComponent />
  <ConnectedReactComponentOwnProps
    label="ConnectedReactComponentOwnProps Label"
  />
</React.Fragment>
`;
